### ISSUE TITLE
Add support for 'XmlnsDefinitionAttribute'

### ISSUE THREAD

**Issue by dotMorten** [Microsoft/xaml-standard#141]
[XmlnsDefinitionAttribute](https://msdn.microsoft.com/en-us/library/system.windows.markup.xmlnsdefinitionattribute(v=vs.110).aspx) is an absolutely great attribute for Control Library developers. It allows the developers of that control library to greatly simplify the xaml by reducing the number of XMLNS declarations they need. 
The only way around this is for the control vendor to throw every single class that might be used in XAML (not just controls but other types too), in the same namespace, creating a very ugly unwieldy API to code against.
If you use a 3rd party control library or you're a control library vendor, you should vote for this. It'll save you a lot of typing, looking for controls and types etc.

WPF and Xamarin.Forms already support this, but it is missing from UWP.

This is already a highly voted issue on UserVoice with several key 3rd party library vendors chiming in as well: https://wpdev.uservoice.com/forums/110705-universal-windows-platform/suggestions/9523650-add-support-for-xmlnsdefinitionattribute

**Comment by insinfo** [comment]
@dotMorten
I did not understand exactly what you want, you could exemplify by writing the syntax
If you can help me in defining the syntax of these controls
[Visit this link](https://github.com/Microsoft/xaml-standard/issues/117),

**Comment by dotMorten** [comment]
I'm not sure what you mean by that. The feature is well documented and understood by the XAML teams so don't see a need to also re-document it here. Also I don't see the relationship of this to the other issue you refer to.

**Comment by birbilis** [comment]
also using it in http://github.com/zoomicon/Compatibility
to map between WPF and Silverlight and be able to share XAML and C# between such projects via linked files. Something like polyfiling

**Comment by dotMorten** [comment]
@birbilis I fail to understand the relevance of your link to xmlns definition attribute? What do you mean?

**Comment by BillHenning** [comment]
I completely agree.  As a control developer, this is one of the single most annoying things about UWP and nobody has ever seemed to listen at Microsoft about it.  It doesn't even seem like it should be that "hard" for them to support it in VS tooling either.  I fail to understand why it was left out in WinRT/UWP.  This most certainly should be in XAML Standard.

**Comment by birbilis** [comment]
I mean I was needing that attribute to implement compatibility layer between WPF and Silverlight, so that I could use common source files as linked files between such projects. So it's use is not just for avoiding the writing of multiple "imports" in your xaml, but also to be able to reuse xaml between different platforms

**Comment by birbilis** [comment]
Oops, meant to write "its"

**Comment by dotMorten** [comment]
@birbilis you can click the little  pen edit icon to fix your typos. Just like you can use the 'x' to delete a comment you didn't mean to (ie I'll be deleting this comment again since it really doesn't belong in this thread, but you should at least get this in your email)

**Comment by Mike-EEE** [comment]
PROFESSOR.  DOT. MORTEN!!! 😆 😆 😆  ❤️

**Comment by birbilis** [comment]
(since I got a comment on this, on windows phone github shows no UI to edit a reply you just did)

**Comment by birbilis** [comment]
regarding the usage I mentioned above, see
https://github.com/Zoomicon/Compatibility/blob/master/Compatibility.Silverlight/Properties/AssemblyInfo.cs
where it writes:

[assembly: XmlnsDefinition("http://schemas.microsoft.com/winfx/2006/xaml/presentation", "System.Windows.Controls")]

//for KeyboardNavigation:
[assembly: XmlnsDefinition("http://schemas.microsoft.com/winfx/2006/xaml/presentation", "System.Windows.Input")]
[assembly: XmlnsDefinition("http://schemas.microsoft.com/client/2007", "System.Windows.Input")]

//for FocusManager:
[assembly: XmlnsDefinition("http://schemas.microsoft.com/client/2007", "Compatibility")]

//for Label, Accordion:
[assembly: XmlnsDefinition("http://schemas.microsoft.com/winfx/2006/xaml/presentation", "Compatibility")]
[assembly: XmlnsDefinition("http://schemas.microsoft.com/winfx/2006/xaml/presentation/sdk", "Compatibility")]

//for BusyIndicator, DockPanel, Accordion, AccordionItem, Expander:
[assembly: XmlnsDefinition("http://schemas.microsoft.com/winfx/2006/xaml/presentation/toolkit", "Compatibility")]
[assembly: XmlnsDefinition("http://schemas.microsoft.com/winfx/2006/xaml/presentation/toolkit/extended", "Compatibility")]`

similar XmlnsDefinition are at the Compatibility.WPF etc. projects (I see the .WP7 one doesn't, probably that one was untested since I needed older Visual Studio to open it) - those projects link shared files from the Source folder and have conditionals in their code so that for example Visual.cs ends up being an empty file for WPF (in such extreme cases could just opt to not include the file in the WPF project, but still I use conditionals to avoid accidental inclusion). 

BTW, would be nice if XAML had support for conditional for parts of the XML tree (like XML comments with some special annotation). Older parsers would just ignore all those and newer ones would check only the parts that match the given conditionals. Could even do that with preparsing btw, at the IDE level I mean with some VS extension. That way I could write XAML that could be added as linked file at multiple projects in a solution

**Comment by lokitoth** [comment]
One issue with this is that it starts going beyond defining a vocabulary and into the runtime implementation detail: By specifying this, you limit the set of possible runtimes which could support XAML Standard.

On the flip side, I don't see how we could define a XAML vocabulary without defining how to "import" namespaces and assemblies, and I was planning on opening up an item around the xmlns:<xyz>="using:<abc>" construction, because it doesn't allow disambiguation (nor does the uri-like namespace support) if two different controls in different control libraries happen to collide, ala: https://github.com/lokitoth/XAMLNamespaceCollision

With that said, this seems like a good place to consider a pure-XAML way of joining up multiple namespaces into a control lib grouped by URI, without relying on details of the platform used for the code-behind.

**Comment by birbilis** [comment]
btw, since I mentioned conditionals in XAML, also see the OnPlatform and the Ignorable at https://github.com/Microsoft/xaml-standard/issues/14. Maybe such conditional syntax could help put the different namespaces (used in scenarios like the WPF/Silverlight compatibility layer I mentioned above) in the XAML itself instead of the code-behind. 

BTW, the other use of XmlnsDefinition, that is the merging of multiple namespaces as one for usage in the XAML is also very important so that your component users don't have to import multiple namespaces to use your component, nor you as component author have to dump all your object/subobject modelling under one flat namespace to not make your users' (and tooling's) life hard
